src/gluonts/dataset/schema.py

```diff
@@ -0,0 +1,267 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Optional, Union, Generic, TypeVar, Type
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from pandas.tseries.frequencies import to_offset
+from torch import layout
+
+from gluonts.exceptions import GluonTSDataError
+
+
+@dataclass
+class PyArray:
+    array_shape: list
+    dtype: Type
+
+    def __str__(self):
+        return "array(shape=%s, dtype=%s)" % (self.array_shape, self.dtype)
+
+
+T = TypeVar("T")
+
+
+class FieldType(Generic[T]):
+    def __call__(self, v: Any) -> T:
+        raise NotImplementedError()
+
+    def __eq__(self, other):
+        raise NotImplementedError()
+
+    def is_compatible(self, v: Any) -> bool:
+        raise NotImplementedError()
+
+
+@dataclass
+class NumpyArrayField(FieldType[np.ndarray]):
+    dtype: Type = np.float32
+    ndim: Optional[int] = None
+
+    def __call__(self, value: Any) -> np.ndarray:
+        if isinstance(value, pa.Array):
+            value = value.to_numpy()
+        else:
+            value = np.asarray(value, dtype=self.dtype)
+        if self.ndim is not None and self.ndim != value.ndim:
+            raise GluonTSDataError(
+                f"expected array with dimension {self.ndim}, "
+                "but got {value.ndim}."
+            )
+
+        return value
+
+    def is_compatible(self, value: Any) -> bool:
+        if not isinstance(value, (list, tuple, np.ndarray, pa.Array)):
+            return False
+
+        try:
+            if isinstance(value, pa.Array):
+                x = value.to_numpy()
+            else:
+                x = np.asarray(value)
+        except Exception:
+            return False
+
+        # int types
+        if self.dtype in [np.int32, np.int64]:
+            if x.dtype.kind != "i":
+                return False
+            return self.ndim is None or x.ndim == self.ndim
+
+        # float types
+        try:
+            x = np.asarray(value, dtype=self.dtype)
+        except Exception:
+            return False
+        return self.ndim is None or x.ndim == self.ndim
+
+
+Freq = Union[str, pd.DateOffset]
+
+
+class PandasPeriodField(FieldType[pd.Period]):
+    def __init__(
+        self,
+        freq: Freq,
+    ) -> None:
+        self.freq = to_offset(freq) if isinstance(freq, str) else freq
```

Comment:
I think you can always call `to_offset` since it just returns the value if it is already an offset.